### PR TITLE
Fixes #279 : adding CI support for sub projects.

### PR DESCRIPTION
**Pull request by sch00lb0y**
## Description

I have added a feature to register sub projects to CI
## Motivation and Context

Right now, user can register only one repository to CI, so webhook are not registered to other repository. so, I have added support to register webhook to all sub repositories. 
## How Has This Been Tested?

https://demo08.herokuapp.com
## Types of changes

- [ ] Bug fix 
- [x] New feature

## Checklist:

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [x] There is only one commit per issue.

### CODE REVIEW

**Comment by sch00lb0y** [comment]
@pri22296 @imujjwal96 please review

File: backend/github.js [hunk 1]
```diff
@@ -72,4 +72,85 @@ exports.hasAdminAccess = function (repository, accessToken, callback) {
...
+};
+/**
+* hookValidator is used to check wheather hook is registered to
+* repository or not
+* @param repositoryName: name of the repository
+* @param token: github access token
+*/
+exports.hookValidator = function (repositoryName, token) {
+  return new Promise(function(resolve, reject) {
+    token = crypter.decrypt(token);
```
Comment (imujjwal96) [inline backend/github.js]: No need to create a separate variable as it is used only once

File: backend/github.js [hunk 2]
```diff
@@ -72,4 +72,85 @@ exports.hasAdminAccess = function (repository, accessToken, callback) {
...
+    request({
+      url: `https://api.github.com/repos/${repositoryName}/hooks?access_token=${token}`,
```
Comment (imujjwal96) [inline backend/github.js]: instead of adding a query parameter for the access token, add it in the authorization header.

File: backend/github.js [hunk 3]
```diff
@@ -72,4 +72,85 @@ exports.hasAdminAccess = function (repository, accessToken, callback) {
...
+      headers: {
+        'User-Agent': 'request'
```
Comment (imujjwal96) [inline backend/github.js]: i'd say we should use `Yaydoc` as the user agent

File: backend/github.js [hunk 4]
```diff
@@ -72,4 +72,85 @@ exports.hasAdminAccess = function (repository, accessToken, callback) {
...
+      }
+    }, function (error, response, body) {
+      var isRegistered = false;
+      if (response.statusCode !== 200) {
+        console.log(response.statusCode + ': ' + response.statusMessage);
+        reject()
+      }
+      var hooks = JSON.parse(body);
+      var hookurl = 'https://' + process.env.HOSTNAME + '/ci/webhook';
```
Comment (imujjwal96) [inline backend/github.js]: do not strict it to https if a deployment does not have SSL encryption the webhooks will fail. Had this problem in my deployment

File: backend/github.js [hunk 5]
```diff
@@ -72,4 +72,85 @@ exports.hasAdminAccess = function (repository, accessToken, callback) {
...
+      hooks.forEach(function (hook) {
+        if (hook.config.url === hookurl) {
+          isRegistered = true;
```
Comment (imujjwal96) [inline backend/github.js]: break it out of loop after the hookurl is equated.

File: backend/github.js [hunk 6]
```diff
@@ -72,4 +72,85 @@ exports.hasAdminAccess = function (repository, accessToken, callback) {
...
+*/
+exports.registerHook = function (data, token) {
+return new Promise(function(resolve, reject) {
+  token = crypter.decrypt(token)
+  var hookurl = 'https://' + process.env.HOSTNAME + '/ci/webhook';
+  if (data.sub === true) {
+    hookurl += `?sub=true`;
+  }
+  request({
+    url: `https://api.github.com/repos/${data.repositoryName}/hooks?access_token=${token}`,
```
Comment (imujjwal96) [inline backend/github.js]: same as before....

File: backend/github.js [hunk 7]
```diff
@@ -72,4 +72,85 @@ exports.hasAdminAccess = function (repository, accessToken, callback) {
     var adminPermission = bodyJSON.permissions.admin || false;
     callback(null, adminPermission);
   });
-};
\ No newline at end of file
+};
+/**
+* hookValidator is used to check wheather hook is registered to
```
Comment (imujjwal96) [inline backend/github.js]: add a single whitespace before *. All * should be in a single line

File: backend/github.js [hunk 8]
```diff
@@ -72,4 +72,85 @@ exports.hasAdminAccess = function (repository, accessToken, callback) {
...
+    }
+  });
+});
+}
+
+/**
+* @param repoName: repository name (fossasia/yaydoc)
+* returns clone url
+*/
+
```
Comment (imujjwal96) [inline backend/github.js]: remove line

File: backend/github.js [hunk 9]
```diff
@@ -72,4 +72,85 @@ exports.hasAdminAccess = function (repository, accessToken, callback) {
...
+    if (response.statusCode !== 201) {
+      resolve(false);
+    } else {
+      resolve(true);
+    }
+  });
+});
+}
+
+/**
```
Comment (imujjwal96) [inline backend/github.js]: add a single line description to maintain consitency

File: backend/github.js [hunk 10]
```diff
@@ -72,4 +72,85 @@ exports.hasAdminAccess = function (repository, accessToken, callback) {
...
+* @param repoName: repository name (fossasia/yaydoc)
+* returns clone url
+*/
+
+exports.getCloneUrl = function (repoName) {
+  var username = repoName.split('/')[0];
```
Comment (imujjwal96) [inline backend/github.js]: what's the need to split them?
username/name is same as reponame

Comment (imujjwal96) [inline backend/github.js]: That  raises the question in the existence of this function

File: public/scripts/socket-dashboard.js [hunk 11]
```diff
@@ -1,46 +1,70 @@
...
  * Search for repositories
  */
-var search = function () {
-  var username = $("#orgs").val().split(":")[1];
-  const searchBarInput = $("#search_bar");
-  const searchResultDiv = $("#search_result");
+var subProjectId = 0; // sub projects dynamic id
+var repositorySelectIds = []; // array for storing repositories select element
+
+function registerDisableChecker () {
```
Comment (imujjwal96) [inline public/scripts/socket-dashboard.js]:
```
var registerDisableChecker = function() {
  ....
  ....
}
```
Maintain consistency

File: public/scripts/socket-dashboard.js [hunk 12]
```diff
@@ -1,46 +1,70 @@
...
+  var flag = true; // flag to check wheather any of the select is
```
Comment (imujjwal96) [inline public/scripts/socket-dashboard.js]: *whether Spell check

File: routes/ci.js [hunk 13]
```diff
@@ -22,154 +23,131 @@ router.post('/register', function (req, res, next) {
...
-
-          });
-        } else {
+      var hookValidation = Promise.all([repositoryName].concat(subRepositories).map(x => github.hookValidator(x, token)))
+      hookValidation.then(function (validatedResults) {
+        if (validatedResults.indexOf(true) > -1) {
           res.redirect("/dashboard?status=registration_already");
-        }
+        } else {
+          var registerdHook = Promise.all([{repositoryName:repositoryName,sub:false}]
```
Comment (imujjwal96) [inline routes/ci.js]: *registeredHook spell check

Comment (sch00lb0y) [inline routes/ci.js]: updated.

**Comment by imujjwal96** [comment]
What will happen when another project is added?

**Comment by sch00lb0y** [comment]
@imujjwal96 could you elaborate your query?

**Comment by imujjwal96** [comment]
Initally, i defined 3 subprojects in my main projects' `.yaydoc.yml`. If I plan to add a new subproject, what will i have to do?

**Comment by sch00lb0y** [comment]
@imujjwal96 actually user have to register ci for sub projects in CI dashboard. I have created a form for adding sub project similar to `WEBUI`. and it'll get saved in DB. if any changes happen to sub project documentation will be generated and pushed to main repository Github pages

**Comment by imujjwal96** [comment]
What if i want to add another sub project?
Consider i have a project with 3 subprojects. 
 - android app
 - rest api
 - web app

I registered them using the WebUI and all's going well. 

Now if i create add an iOS application and wish to add it as a subproject. What will happen then? How will i register that as a subproject for my master project?

**Comment by sch00lb0y** [comment]
right now we have only option to enable and disable CI. we don't have the option to edit CI config. please raise as a separate issue I'll work on it

**Comment by imujjwal96** [comment]
@sch00lb0y Let me see and work on it :)

**Comment by sch00lb0y** [comment]
@pri22296 @imujjwal96 I resolved conflicts please check. https://demo08.herokuapp.com

**Comment by sch00lb0y** [comment]
@imujjwal96 @pri22296 please check this

**Comment by pri22296** [comment]
@sch00lb0y Update with option to add another subproject after registering?

**Comment by imujjwal96** [comment]
@pri22296 Not in this PR. Things will only become difficult for us to review.

**Comment by sch00lb0y** [comment]
@pri22296 @imujjwal96 I resolved conflicts. Please review

File: backend/github.js [hunk 14]
```diff
@@ -96,3 +96,75 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
...
+        'Authorization': 'token ' + crypter.decrypt(token)
+      }
+    }, function (error, response, body) {
+      var isRegistered = false;
+      if (response.statusCode !== 200) {
+        console.log(response.statusCode + ': ' + response.statusMessage);
+        reject()
+      }
+      var hooks = JSON.parse(body);
+      var hookurl = 'http://' + process.env.HOSTNAME + '/ci/webhook';
```
Comment (pri22296) [inline backend/github.js]: Would using `http` be a problem? I'm not sure, jusk asking.

File: backend/github.js [hunk 15]
```diff
@@ -96,3 +96,75 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
...
+}
+
+/**
+ * registerHook is used to register hook to the respository
+ * @param repositoryName: name of the repository
+ * @param token: github access token
+ */
+exports.registerHook = function (data, token) {
+return new Promise(function(resolve, reject) {
+  var hookurl = 'http://' + process.env.HOSTNAME + '/ci/webhook';
```
Comment (pri22296) [inline backend/github.js]: Here too

File: public/scripts/socket-dashboard.js [hunk 16]
```diff
@@ -1,46 +1,70 @@
...
  */
-var search = function () {
-  var username = $("#orgs").val().split(":")[1];
-  const searchBarInput = $("#search_bar");
-  const searchResultDiv = $("#search_result");
+var subProjectId = 0; // sub projects dynamic id
+var repositorySelectIds = []; // array for storing repositories select element
+
+var registerDisableChecker = function () {
+  var flag = true; // flag to check wheather any of the select is
```
Comment (pri22296) [inline public/scripts/socket-dashboard.js]: typo: _wheather_ -> _whether_

.. any of the select is ..? This sentence looks incomplete.

File: public/scripts/socket-dashboard.js [hunk 17]
```diff
@@ -113,4 +126,35 @@ $(function() {
...
+      search($(searchBar), $(searchResult), id);
+    });
+  }
+
+  registerSearchEvents(0);
+
+  $("#btnAddSub").click(function () {
+    subProjectId += 1;
+    styles.disableButton("btnRegister");
+    $("#SubProject").append(`<div class="form-group" class="subProject_${subProjectId}"><label for="search_bar" class="control-label">Sub Project ${subProjectId}:</label><div class="input-group"><input id="search_bar_${subProjectId}" placeholder="Enter project name" class="form-control"><span class="input-group-btn"><button id="search_${subProjectId}" type="button" class="btn btn-default">Search</button></span></div></div><div id="search_result_${subProjectId}"></div>`);
```
Comment (pri22296) [inline public/scripts/socket-dashboard.js]: Can this line be broken into multiple lines?

File: backend/github.js (hunk 14 above)
Comment (sch00lb0y) [inline backend/github.js]: >do not strict it to https if a deployment does not have SSL encryption the webhooks will fail. Had this problem in my deployment

Comment (pri22296) [inline backend/github.js]: Can't we use the current protocol of the browser by using just `//`? Like if the site is opened using http, use `http` else use `https`

File: public/scripts/socket-dashboard.js (hunk 16 above)
Comment (sch00lb0y) [inline public/scripts/socket-dashboard.js]: updated

File: public/scripts/socket-dashboard.js (hunk 17 above)
Comment (sch00lb0y) [inline public/scripts/socket-dashboard.js]: updated

File: backend/github.js (hunk 14 above)
Comment (sch00lb0y) [inline backend/github.js]: hookurl is accessed by GitHub. Once we done with ssl in kuernetes we'll change this

**Comment by imujjwal96** [comment]
This does not seem to be working. I register BrimeNotes/documentation as the master project with BrimeNotes/core and BrimeNotes/desktop as the subprojects. When I made a commit to BrimeNotes/core, the documentation is not generated.

**Comment by sch00lb0y** [comment]
@imujjwal96 I have checked. it is working. I have cleared the db now. could you remove the webhook manually and check once again ?

**Comment by imujjwal96** [comment]
Not working!!

**Comment by sch00lb0y** [comment]
@imujjwal96 I found why app is crashing. I'm querying owner.id instead of registrant.id that's why app is crashing for the organization. now I have fixed now you make changes is documentation it'll work no need for re registrarion

**Comment by imujjwal96** [comment]
Not able to register apps.
image

If this is because of empty subprojects, then please handle it accordingly. User would not like to refresh the page if an extra form element for a subproject is added

**Comment by imujjwal96** [comment]
@pri22296 @sch00lb0y  This PR is generating docs for sub projects defined in Yaydoc CI, overriding .yaydoc.yml. Should this be expected? And if registering sub projects here is enough, why do we need them in the yaml file

**Comment by pri22296** [comment]
@imujjwal96 Yeah, when this feature is complete, I think subprojects can be removed from the yaml file.
By complete I mean, option to delete, modify, add new subproject.

**Comment by imujjwal96** [comment]
@pri22296 What if instead of registering sub projects at YaydocCI, we grab them from the yaydoc.yml and then register?

**Comment by pri22296** [comment]
Yeah, that could work. But how will changes to the yaml file be handled? And any advantage for that approach compared to the current implementation? Coz I think it, that it would take significant amount of work, would it not?

**Comment by imujjwal96** [comment]
What decides the source for subproject in the current implementation?

**Comment by imujjwal96** [comment]
What we can do is.
* Add `.yaydoc.yml` in a master project
* Register the project to Yaydoc CI 
* On registering, the yaml file is checked for subprojects. If subprojects are defined, we register these projects and their hooks.
* Thereafter on each commit we check if a change is made in yaydoc.yml (maybe store a key in repository database (subprojects: true)). If the change is made, we check if subprojects are added or removed. On this change, we add or remove hooks for those repositories

**Comment by sch00lb0y** [comment]
how will we handle if the user doesn't have access to sub project? if we implement this we have to change for everything like reading theme, doc path for normal project generation as well.  will it be possible for us to change the whole implementation within the small time span and make the documentation for all FOSSASIA project up and running ?

**Comment by imujjwal96** [comment]
* What decides the source for subprojects in the current implementation?
* What decides the branch for subprojects in the current implementation?

**Comment by imujjwal96** [comment]
Moving forward with the current implementation, when i register BrimeNotes/documentation and imujjwal96/brimeAPI, i get application error

**Comment by sch00lb0y** [comment]
I have checked the condition. it is working for me. I have cleared the db could you remove the webhook and try again ?

**Comment by sch00lb0y** [comment]
@imujjwal96 I have checked by forking your repository. it is working for me. please confirm for me

File: backend/github.js [hunk 18]
```diff
@@ -96,3 +96,75 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
     }
   });
 };
+
+/*
+ * hookValidator is used to check whether hook is registered to repository or not
```
Comment (imujjwal96) [inline backend/github.js]: *Check whether hook is registered to repository or not

File: backend/github.js [hunk 19]
```diff
@@ -96,3 +96,75 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
...
+ * @param repositoryName: name of the repository
```
Comment (imujjwal96) [inline backend/github.js]: @param name: `full_name` of the repository
Follow the style used while commenting other functions in the project

File: backend/github.js [hunk 20]
```diff
@@ -96,3 +96,75 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
...
+      ],
+      config: {
+        url: hookurl,
+        content_type: "json"
+      }
+    }
+  }, function(error, response, body) {
+    if (response.statusCode !== 201) {
+      console.log(response.statusCode + ': ' + response.statusMessage);
+      resolve({status: false, body:body});
```
Comment (imujjwal96) [inline backend/github.js]: Why haven't you used reject here?

File: backend/github.js [hunk 21]
```diff
@@ -96,3 +96,75 @@ exports.deleteHook = function (name, hook, accessToken, callback) {
...
+          break;
+        }
+      }
+      resolve({repositoryName: repositoryName, isRegistered: isRegistered})
+    });
+  });
+}
+
+/**
+ * registerHook is used to register hook to the respository
```
Comment (imujjwal96) [inline backend/github.js]: Follow practices already used in the project

File: public/scripts/socket-dashboard.js [hunk 22]
```diff
@@ -113,4 +126,44 @@ $(function() {
       $("#disableModal").modal('hide');
     });
   });
+
+  var registerSearchEvents = function (id) {
```
Comment (imujjwal96) [inline public/scripts/socket-dashboard.js]: add documentation

File: backend/github.js (hunk 20 above)
Comment (sch00lb0y) [inline backend/github.js]: `Promise.all` will go to catch if we reject. we have to remove the previously registered webhook if any of the webhook fails. I didn't write the function but for the future, I wrote

Comment (imujjwal96) [inline backend/github.js]: Can we not remove the registered hook from catch?

Comment (sch00lb0y) [inline backend/github.js]: catch won't return the array of the previously registered value. it just returns the value where catch has been executed

**Comment by sch00lb0y** [comment]
@imujjwal96 updated.

**Pull request merged**